### ISSUE TITLE
Infinite loop between Jenkins instance and Bitbucket

### ISSUE THREAD

**Issue by joinghel** [mallowlabs/bitbucket-oauth-plugin#18]
Hi,

I just upgraded from v0.7 to v0.8 and followed the migration instructions. After restarting Jenkins, I'm continuously redirected between Bitbucket and Jenkins (see screenshot). The only thing I changed was the callback url. I verified if the client_id and secret were still ok and they are.

I don't get any errors in the Jenkins log file.

Any idea how I can solve this? If you need more details, let me know and I'll provide them.

Thanks!

![image](https://user-images.githubusercontent.com/22074706/51314510-23b0b580-1a50-11e9-857b-2e74bef01c03.png)

**Comment by mallowlabs** [comment]
@joinghel 

I guess that Jenkins security matrix is misconfigured.

Please check `Logged-in users can do anything` in Authorization setting.
Can you log-into Jenkins with Bitbucket OAuth Plugin?

<img width="395" alt="authorization" src="https://user-images.githubusercontent.com/19996/51318948-17077e00-1a9f-11e9-9941-972f5b321091.png">

**Comment by joinghel** [comment]
My current config is as follows:

```
<useSecurity>true</useSecurity>
<authorizationStrategy class="hudson.security.FullControlOnceLoggedInAuthorizationStrategy">
  <denyAnonymousReadAccess>true</denyAnonymousReadAccess>
</authorizationStrategy>
<securityRealm class="org.jenkinsci.plugins.BitbucketSecurityRealm">
  <clientID>****</clientID>
  <clientSecret>****</clientSecret>
</securityRealm>
<disableRememberMe>false</disableRememberMe>
```

So that seems to be correct. When I manually call `https://our.jenkins/securityRealm/finishLogin?code=sometestcode`, I get the following exception:

```
org.scribe.exceptions.OAuthException: Cannot extract an acces token. Response was: {"error_description": "The specified code is not valid.", "error": "invalid_grant"}
```

I don't get this exception when trying to log in, so that means the code must be valid. So I guess that for some reason, I'm just nog logged in at the end of the finishLogin call. Is this plugin responsible for the login?

I tried in incognito, different browser, using another device.

I also read the other issues in the repo and updated our jenkins instance, auto-removed old kernel headers, rebooted the server, restarted jenkins multiple times, etc.

**Comment by mallowlabs** [comment]
Tell me your Jenkins version.

**Comment by joinghel** [comment]
2.150.2
On Ubuntu 16.04.5 LTS

**Comment by joinghel** [comment]
When running `apt-get upgrade jenkins` I get the message that it is the latest version. However, [here](https://jenkins.io/changelog/), I see that I'm way behind. I'm using https://pkg.jenkins.io/debian-stable .

**Comment by mallowlabs** [comment]
on 2.150.2, I got infinite redirect loop too!
It looks like a bug. I investigate this problem.

**Comment by jrnt30** [comment]
Wow! This happened when I was changing some ingress rules and I thought somehow I had caused this.  Somewhat nice to know that I'm not alone! 

A few things we saw were that the:
- Role endpoints are being properly called to assess the groups in the plugin (via logs) 
- the `/whoAmI` endpoint was returning the users as anonymous and not mapping to any privs

**Comment by mallowlabs** [comment]
I probably fixed redirect loop problem on f7ba72f.
I will test this commit and release to jenkins plugin center in this weekend.
Thank you for your patience.

**Comment by joinghel** [comment]
That's great, thanks! I'll let you know if it works on our side.

Thank you for the fast response & solution!

**Comment by moimart1** [comment]
Duplicate issue here: https://issues.jenkins-ci.org/browse/JENKINS-55668
With a similar PR https://github.com/jenkinsci/bitbucket-oauth-plugin/pull/3

**Comment by mackoftrack** [comment]
ETA for a fix? This broke jenkins for us

**Comment by mallowlabs** [comment]
I merged https://github.com/jenkinsci/bitbucket-oauth-plugin/pull/3 .
I released v0.9.
Please update bitbucket-oauth-plugin to v0.9.

**Comment by mackoftrack** [comment]
Thanks! It works!
